File: day17.pl
#!/usr/bin/env perl 
#===============================================================================
#
#         FILE: day17.pl
#
#        USAGE: ./day17.pl  
#
#  DESCRIPTION: Day 17 - Packaging eggnog for the frig.
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: Mark T. Johnson (mtj), markj@matzsoft.com
# ORGANIZATION: MATZ Software & Consulting
#      VERSION: 1.0
#      CREATED: 12/16/15 23:59:13
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.010;

use List::Util qw( min );


my $eggnog = 150;
my @containers;

while ( <DATA> ) {
    chomp;

    push @containers, $_;
}

my @solutions = find_solutions( 0, $eggnog, @containers );

say scalar( @solutions );

my $minimum = min @solutions;

say scalar( grep { $_ == $minimum } @solutions );



sub find_solutions {
    my $containers = shift;
    my $quantity   = shift;
    my $first      = shift;
    my @rest       = @_;

    if ( @rest == 0 ) {
        return $first == $quantity ? ( $containers + 1 ) : ();
    }

    if ( $first > $quantity ) {
        return find_solutions( $containers, $quantity, @rest );

    } elsif ( $first == $quantity ) {
        return ( $containers + 1,
            find_solutions( $containers, $quantity, @rest ) );

    } else {
        return (
            find_solutions( $containers + 1, $quantity - $first, @rest ),
            find_solutions( $containers,     $quantity,          @rest )
        );
    }
}



__END__
33
14
18
20
45
35
16
35
1
13
18
13
50
44
48
6
24
41
30
42

File: 2015/day17.swift
//
//         FILE: main.swift
//  DESCRIPTION: day17 - No Such Thing as Too Much
//        NOTES: ---
//       AUTHOR: Mark T. Johnson, markj@matzsoft.com
//    COPYRIGHT: © 2021 MATZ Software & Consulting. All rights reserved.
//      VERSION: 1.0
//      CREATED: 07/09/21 12:58:36
//

import Foundation

func findSolutions( used: Int, amount: Int, available: [Int] ) -> [Int] {
    let first = available.first!
    let rest = Array( available.dropFirst() )
    
    if rest.isEmpty { return first == amount ? [ used + 1 ] : [] }
    if first > amount { return findSolutions( used: used, amount: amount, available: rest ) }
    if first == amount { return [ used + 1 ] + findSolutions( used: used, amount: amount, available: rest ) }
    
    return
        findSolutions( used: used + 1, amount: amount - first, available: rest ) +
        findSolutions( used: used,     amount: amount,         available: rest )
}


func parse( input: AOCinput ) -> [Int] {
    return input.lines.map { Int( $0 )! }
}


func part1( input: AOCinput ) -> String {
    let containers = parse( input: input )
    let amount = Int( input.extras[0] )!
    let solutions = findSolutions( used: 0, amount: amount, available: containers )
    
    return "\( solutions.count )"
}


func part2( input: AOCinput ) -> String {
    let containers = parse( input: input )
    let amount = Int( input.extras[0] )!
    let solutions = findSolutions( used: 0, amount: amount, available: containers )
    let least = solutions.min()!
    
    return "\( solutions.filter { $0 == least }.count )"
}


try runTests( part1: part1 )
try runTests( part2: part2 )
try solve( part1: part1 )
try solve( part2: part2 )
